Add Fan Speed number entity to Nexia

Ongion opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new number entity to Nexia to support changing the air-handler / blower fan speed on systems with variable-speed fan support.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28627

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Ongion committed:
Add Fan Speed support to Nexia integration

Adds a new "set_fan_speed" service to the Nexia integration, to allow
setting speed of the air-handler/blower fans.

Ongion committed:
Add Fan Speed to Nexia Tests

Ongion commented:
I'm more than happy to add the new service to the Nexia documentation in home-assistant.io, but I wasn't sure on the timing of when to do that. I was expecting to update the documentation once the change was approved, since I didn't want to presume it would be accepted 😅.

gjohansson-ST commented:
> I'm more than happy to add the new service to the Nexia documentation in home-assistant.io, but I wasn't sure on the timing of when to do that. I was expecting to update the documentation once the change was approved, since I didn't want to presume it would be accepted 😅.

Should be done and attached to this PR, see the PR description

Ongion commented:
> Should be done and attached to this PR, see the PR description

Is there a way to mark the documentation PR as being associated with this PR?

MartinHjelmare commented on homeassistant/components/nexia/climate.py:
```diff
@@ -136,6 +139,11 @@ async def async_setup_entry(
         f"async_{SERVICE_SET_HVAC_RUN_MODE}",
     )
+    platform.async_register_entity_service(
+        SERVICE_SET_FAN_SPEED,
```
Why don't add a fan entity instead?

gjohansson-ST replied:
A `Number` entity seems like the middle way here then and not a custom service.

Ongion replied:
That makes sense to me. I can make that change!

Ongion replied:
Changes implemented!

Ongion commented:
Documentation PR added!
https://github.com/home-assistant/home-assistant.io/pull/28627

Ongion committed:
Remove mistakenly-added changes to Climate tests

A previous version of this change made modifications to the base
Climate entity, but that approach was later abandonded. These changes
to Climate tests were left over from that, and should never have been
included.

gjohansson-ST commented:
As already mentioned this should be a `Number` entity so those changes in `climate` and `sensor` would not be needed for this feature. 
I'm again setting this PR to draft while it's reworked. Thanks

Ongion committed:
Add Fan Speed Number Entity

Ongion committed:
Remove Set Fan Speed Service

Ongion committed:
Remove fan_speed attribute

The fan_speed attribute is not necessary with the new Number entity.

bdraco commented on tests/components/nexia/test_number.py:
```diff
@@ -0,0 +1,38 @@
+from .util import async_init_integration
+
+
+async def test_create_fan_speed_sensors(hass: HomeAssistant) -> None:
+    """Test creation of fan speed sensors."""
```
```suggestion
async def test_create_fan_speed_number_entities(hass: HomeAssistant) -> None:
    """Test creation of fan speed number entities."""
```

bdraco commented on homeassistant/components/nexia/number.py:
```diff
@@ -0,0 +1,71 @@
+    coordinator: NexiaDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    nexia_home = coordinator.nexia_home
+
+    entities = []
```
missing typing

bdraco commented on homeassistant/components/nexia/const.py:
```diff
@@ -7,6 +7,7 @@
     Platform.CLIMATE,
     Platform.SCENE,
     Platform.SWITCH,
+    Platform.NUMBER,
```
Please sort these

bdraco commented:
Tested 👍 

<img width="555" alt="Screenshot 2023-08-18 at 10 04 30 PM" src="https://github.com/home-assistant/core/assets/663432/29d34596-8e63-4b26-86a2-d000f7a3b7a4">

Ongion committed:
Address reviewer feedback

Rename test function to reflect fact that fan speed entities are
entities, and not sensors.
Added missing typing to variables.
Sorted list of platforms

gjohansson-ST commented on tests/components/nexia/test_number.py:
```diff
@@ -0,0 +1,38 @@
+from .util import async_init_integration
+
+
+async def test_create_fan_speed_number_entities(hass: HomeAssistant) -> None:
```
Need an additional test that sets the value

gjohansson-ST reviewed (changes requested):
Needs an additional test for coverage otherwise I think this PR is complete

Ongion committed:
Add test_set_fan_speed

Also adds new test fixture for mock response to API call

gjohansson-ST commented on homeassistant/components/nexia/number.py:
```diff
@@ -0,0 +1,72 @@
+        super().__init__(
+            coordinator,
+            thermostat,
+            name=f"{thermostat.get_name()} Fan Speed",
```
```suggestion
            name=f"{thermostat.get_name()} Fan speed",
```

gjohansson-ST committed:
Update homeassistant/components/nexia/number.py

gjohansson-ST reviewed (approved):
Thanks @Ongion 👍

gjohansson-ST commented on tests/components/nexia/test_number.py:
```diff
@@ -0,0 +1,62 @@
+    assert state.state == "35.0"
+    expected_attributes = {
+        "attribution": "Data provided by Trane Technologies",
+        "friendly_name": "Master Suite Fan Speed",
```
```suggestion
        "friendly_name": "Master Suite Fan speed",
```

gjohansson-ST commented on tests/components/nexia/test_number.py:
```diff
@@ -0,0 +1,62 @@
+    assert state.state == "35.0"
+    expected_attributes = {
+        "attribution": "Data provided by Trane Technologies",
+        "friendly_name": "Downstairs East Wing Fan Speed",
```
```suggestion
        "friendly_name": "Downstairs East Wing Fan speed",
```

gjohansson-ST committed:
Name change
